### DISCUSSION TITLE
Consider exposing managers through interfaces

### DISCUSSION THREAD

**Discussion by mwalk-dev** [FirstGearGames/FishNet#654] [Feature Request]
It would be easier to unit test code written against FishNet if the managers were exposed as interfaces to allow for easy mocking.

**Comment by FirstGearGames** [comment]
Can you provide some examples please?

**Comment by mwalk-dev** [comment]
I find myself writing a lot of game logic that wraps/sits on top of core FishNet functionality and I'd like to unit test my functionality and mock everything that isn't the system under test. Most C# mocking frameworks (Moq, NSubstitute, etc) do mocking by implementing an interface or extending a class, and the interface approach is simpler and more robust.

So for example, I have a class that manages clients reconnecting. I want it to do things like:

- Keep track of the player GUID that a pawn belongs to, and remember it even if the client disconnects
- Allow the client X time to reconnect and reassign ownership of the pawn if they reconnect in that time
- If they don't reconnect, despawn the pawn and open that 'slot' back up to matchmaking

This requires building on top of things like `ServerManager.OnRemoteConnectionState`. If `ServerManager` implemented an `IServerManager` interface that exposed the public API, then I could reference that interface instead of the concrete class in all of my game code. That allows me to write unit tests where Moq mocks the `IServerManager` and my unit tests trigger the necessary events in the mock and verify the results. That results in much smaller, cleaner tests with fewer extraneous variables than if I have to actually start a server and clients within each test case.

**Comment by FirstGearGames** [comment]
This is a good idea but it would sit pretty low on the priority list, and with how many other things need to be done probably won't get implemented.

**Comment by FirstGearGames** [comment]
I know it's been awhile since this suggestion has been covered. I do like the idea and I've been trying to find time, but there's no way this feature is going to make it in due to work and time restraints. While I support the idea, I'm going to have to close this out as declined.
